src/shared_memory_manager.cc

```diff
@@ -60,72 +238,202 @@ SharedMemoryManager::GetCUDAHandle(
 {
   return TRITONSERVER_ErrorNew(
       TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
+      std::string(
+          "GPU shared memory features are currently not supported on Windows")
           .c_str());
 }
+
 #endif  // TRITON_ENABLE_GPU
 
 TRITONSERVER_Error*
 SharedMemoryManager::GetMemoryInfo(
     const std::string& name, size_t offset, void** shm_mapped_addr,
     TRITONSERVER_MemoryType* memory_type, int64_t* device_id)
 {
-  return TRITONSERVER_ErrorNew(
-      TRITONSERVER_ERROR_UNSUPPORTED,
-      std::string("Shared memory feature is currently not supported on Windows")
-          .c_str());
+  // protect shared_memory_map_ from concurrent access
+  std::lock_guard<std::mutex> lock(mu_);
+
+  auto it = shared_memory_map_.find(name);
+  if (it == shared_memory_map_.end()) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_NOT_FOUND,
+        std::string("Unable to find shared memory region: '" + name + "'")
+            .c_str());
+  }
+
+  // validate offset
+  size_t max_offset = 0;
+  if (it->second->kind_ == TRITONSERVER_MEMORY_CPU) {
+    max_offset = it->second->offset_;
+  }
+  if (it->second->byte_size_ > 0) {
+    max_offset += it->second->byte_size_ - 1;
+  }
+  if (offset > max_offset) {
+    return TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INVALID_ARG,
+        std::string("Invalid offset for shared memory region: '" + name + "'")
+            .c_str());
+  }
+  // TODO: should also validate byte_size from caller
+
+  if (it->second->kind_ == TRITONSERVER_MEMORY_CPU) {
+    *shm_mapped_addr = (void*)((uint8_t*)it->second->mapped_addr_ +
+                               it->second->offset_ + offset);
+  } else {
+    *shm_mapped_addr = (void*)((uint8_t*)it->second->mapped_addr_ + offset);
+  }
+
+  *memory_type = it->second->kind_;
+  *device_id = it->second->device_id_;
+
+  return nullptr;
 }
 
 TRITONSERVER_Error*
 SharedMemoryManager::GetStatus(
```

Comment:
Does this need to be implemented twice? I understand the iteration cycle you went though, that's fine to make sure it works, but there is no Windows specific code in this function.

Reply:
The current file structure has duplicated functions for Windows and Unix, with the Windows function simply returning a "not supported error".

Originally, I kept this structure for simplicity. However, as I continued developing, I found it more legible to have separate file segments for Windows/Unix, rather than inlining the differences with ifdefs. However, I understand the software bloat issue here and received similar feedback in the client PR, so I'll try to merge as much together as I can here as well.